// @stackbilt/scaffold-core/dist/__tests__/types.test.js
"use strict";
/**
 * types.test.ts — structural type tests for @stackbilt/scaffold-core
 *
 * Verifies that the package exports the expected symbols and that the
 * TypeScript types compile correctly. Does not invoke any stub implementations.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const index_1 = require("../index");
// ============================================================================
// Function exports
// ============================================================================
(0, vitest_1.describe)('scaffold-core function exports', () => {
    (0, vitest_1.it)('buildScaffold is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.buildScaffold).toBe('function');
    });
    (0, vitest_1.it)('classify is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.classify).toBe('function');
    });
    (0, vitest_1.it)('getKnowledge is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.getKnowledge).toBe('function');
    });
    (0, vitest_1.it)('buildGovernance is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.buildGovernance).toBe('function');
    });
    (0, vitest_1.it)('generateFiles is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.generateFiles).toBe('function');
    });
    (0, vitest_1.it)('materializeScaffold is exported and is a function', () => {
        (0, vitest_1.expect)(typeof index_1.materializeScaffold).toBe('function');
    });
});
// ============================================================================
// Real implementations (classify and buildScaffold are now real)
// ============================================================================
(0, vitest_1.describe)('classify and buildScaffold are implemented', () => {
    (0, vitest_1.it)('buildScaffold returns a LocalScaffoldResult without throwing', () => {
        const result = (0, index_1.buildScaffold)('build a KV-backed worker');
        (0, vitest_1.expect)(result).toBeDefined();
        (0, vitest_1.expect)(result.classification).toBeDefined();
        (0, vitest_1.expect)(result.files).toBeDefined();
    });
    (0, vitest_1.it)('buildScaffold promotes traits and tier2Recommended to top-level', () => {
        const result = (0, index_1.buildScaffold)('multi-tenant SaaS API with JWT auth');
        (0, vitest_1.expect)(Array.isArray(result.traits)).toBe(true);
        (0, vitest_1.expect)(typeof result.tier2Recommended).toBe('boolean');
        (0, vitest_1.expect)(result.traits).toEqual(result.classification.traits);
    });
    (0, vitest_1.it)('workers-saas pattern is assignable as PatternName', () => {
        const p = 'workers-saas';
        (0, vitest_1.expect)(p).toBe('workers-saas');
    });
    (0, vitest_1.it)('classify returns a ClassifyResult without throwing', () => {
        const result = (0, index_1.classify)('build a KV-backed worker');
        (0, vitest_1.expect)(result).toBeDefined();
        (0, vitest_1.expect)(result.pattern).toBeDefined();
        (0, vitest_1.expect)(typeof result.confidence).toBe('number');
    });
});
// ============================================================================
// Type compile tests (satisfies)
// ============================================================================
(0, vitest_1.describe)('types compile', () => {
    (0, vitest_1.it)('PatternName union is assignable', () => {
        const p = 'worker';
        (0, vitest_1.expect)(p).toBe('worker');
    });
    (0, vitest_1.it)('QualityProfile satisfies shape', () => {
        const q = {
            testingLevel: 'standard',
            observability: false,
            authentication: false,
            rateLimiting: false,
            piiHandling: false,
            complianceDomains: [],
        };
        (0, vitest_1.expect)(q.testingLevel).toBe('standard');
    });
    (0, vitest_1.it)('ScaffoldBinding satisfies shape', () => {
        const b = { type: 'KV', name: 'MY_KV', binding: 'MY_KV' };
        (0, vitest_1.expect)(b.type).toBe('KV');
    });
    (0, vitest_1.it)('ThreatEntry satisfies shape', () => {
        const t = {
            id: 'T001',
            category: 'injection',
            description: 'SQL injection via untrusted input',
            mitigation: 'Parameterize all queries',
            severity: 'HIGH',
        };
        (0, vitest_1.expect)(t.severity).toBe('HIGH');
    });
    (0, vitest_1.it)('ScaffoldFile satisfies shape', () => {
        const f = { path: 'src/index.ts', content: '// stub', role: 'entry' };
        (0, vitest_1.expect)(f.role).toBe('entry');
    });
    (0, vitest_1.it)('FileRole union covers expected values', () => {
        const roles = ['entry', 'config', 'test', 'migration', 'contract', 'adf', 'readme'];
        (0, vitest_1.expect)(roles).toHaveLength(7);
    });
    (0, vitest_1.it)('PatternDef satisfies shape', () => {
        const def = {
            name: 'api',
            status: 'ACTIVE',
            category: 'COMPUTE',
            keywords: ['rest', 'http'],
            traits: ['auth'],
        };
        (0, vitest_1.expect)(def.status).toBe('ACTIVE');
    });
    (0, vitest_1.it)('ScaffoldOptions is optional-only', () => {
        const opts = {};
        (0, vitest_1.expect)(opts).toBeDefined();
        const opts2 = { projectName: 'my-worker', oracle: false };
        (0, vitest_1.expect)(opts2.projectName).toBe('my-worker');
    });
    (0, vitest_1.it)('PatternKnowledge satisfies shape', () => {
        const k = {
            threats: [],
            adrContext: '',
            adrDecision: '',
            domainThreats: [],
        };
        (0, vitest_1.expect)(Array.isArray(k.threats)).toBe(true);
    });
    (0, vitest_1.it)('GovernanceDocs satisfies shape (adr002 optional)', () => {
        const g = { threatModel: '', adr001: '', testPlan: '' };
        (0, vitest_1.expect)(g.adr002).toBeUndefined();
    });
    (0, vitest_1.it)('MaterializerResult satisfies shape', () => {
        const facts = {
            pattern: 'worker',
            projectName: 'test',
            intention: 'build a worker',
            bindings: [],
            traits: [],
            qualityProfile: {
                testingLevel: 'basic',
                observability: false,
                authentication: false,
                rateLimiting: false,
                piiHandling: false,
                complianceDomains: [],
            },
        };
        const r = { files: [], facts };
        (0, vitest_1.expect)(r.files).toHaveLength(0);
    });
});
// ============================================================================
// getKnowledge returns empty-but-valid result (knowledge stubs return defaults)
// ============================================================================
(0, vitest_1.describe)('knowledge module stubs return valid empty results', () => {
    (0, vitest_1.it)('getKnowledge returns PatternKnowledge with empty arrays', () => {
        const k = (0, index_1.getKnowledge)('worker');
        (0, vitest_1.expect)(Array.isArray(k.threats)).toBe(true);
        (0, vitest_1.expect)(Array.isArray(k.domainThreats)).toBe(true);
        (0, vitest_1.expect)(typeof k.adrContext).toBe('string');
        (0, vitest_1.expect)(typeof k.adrDecision).toBe('string');
    });
});
