// @stackbilt/core/dist/schemas.js
"use strict";
/**
 * Zod validation schemas for Charter governance data.
 *
 * These schemas define the shape of governance artifacts (ledger entries,
 * patterns, protocols, requests) and can be used for input validation
 * in both Kit CLI and Cloud API contexts.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateProjectSchema = exports.ResolveGovernanceRequestSchema = exports.CreateGovernanceRequestSchema = exports.CreateProtocolSchema = exports.UpdatePatternSchema = exports.CreatePatternSchema = exports.UpdateLedgerStatusSchema = exports.CreateLedgerEntrySchema = void 0;
const zod_1 = require("zod");
// ============================================================================
// Ledger Entry Schemas
// ============================================================================
exports.CreateLedgerEntrySchema = zod_1.z.object({
    entry_type: zod_1.z.enum(['RULING', 'ADR', 'POLICY', 'SOP', 'STRATEGIC', 'REVIEW', 'NOTARY_STAMP']),
    source_mode: zod_1.z.enum(['PRODUCT', 'UX', 'RISK', 'ARCHITECT', 'TDD', 'SPRINT']),
    title: zod_1.z.string()
        .min(5, "Title must be at least 5 characters")
        .max(100, "Title must be at most 100 characters"),
    summary: zod_1.z.string().max(500).optional(),
    input_excerpt: zod_1.z.string().max(500).optional(),
    output: zod_1.z.string().min(10, "Output must be at least 10 characters"),
    tags: zod_1.z.array(zod_1.z.string()).optional(),
    project_id: zod_1.z.string().optional(),
    artifact_hash: zod_1.z.string().regex(/^sha256:[A-Fa-f0-9]{64}$/).optional(),
    quality_metadata: zod_1.z.object({
        specificity_score: zod_1.z.number().min(0).max(100),
        rubric_version: zod_1.z.string().min(1)
    }).optional(),
    stamp: zod_1.z.object({
        stamp_id: zod_1.z.string().min(1),
        ledger_ids: zod_1.z.array(zod_1.z.string().uuid()).min(1),
        issued_at: zod_1.z.string().min(1),
        policy_hash: zod_1.z.string().regex(/^sha256:[A-Fa-f0-9]{64}$/),
        signature: zod_1.z.string().min(8)
    }).optional(),
    create_pattern: zod_1.z.object({
        name: zod_1.z.string().min(3).max(100),
        category: zod_1.z.enum(['COMPUTE', 'DATA', 'INTEGRATION', 'SECURITY', 'ASYNC']),
        blessed_solution: zod_1.z.string().min(5).max(200),
        rationale: zod_1.z.string().max(1000).optional(),
        anti_patterns: zod_1.z.string().max(500).optional(),
        documentation_url: zod_1.z.string().url().optional()
    }).optional()
}).superRefine((data, ctx) => {
    if (data.entry_type === 'NOTARY_STAMP') {
        if (!data.artifact_hash) {
            ctx.addIssue({ code: zod_1.z.ZodIssueCode.custom, message: 'artifact_hash is required for NOTARY_STAMP' });
        }
        if (!data.quality_metadata) {
            ctx.addIssue({ code: zod_1.z.ZodIssueCode.custom, message: 'quality_metadata is required for NOTARY_STAMP' });
        }
        if (!data.stamp) {
            ctx.addIssue({ code: zod_1.z.ZodIssueCode.custom, message: 'stamp is required for NOTARY_STAMP' });
        }
    }
});
exports.UpdateLedgerStatusSchema = zod_1.z.object({
    status: zod_1.z.enum(['ACTIVE', 'SUPERSEDED', 'ARCHIVED'])
});
// ============================================================================
// Pattern Schemas
// ============================================================================
exports.CreatePatternSchema = zod_1.z.object({
    name: zod_1.z.string().min(3).max(100),
    category: zod_1.z.enum(['COMPUTE', 'DATA', 'INTEGRATION', 'SECURITY', 'ASYNC']),
    blessed_solution: zod_1.z.string().min(5).max(200),
    rationale: zod_1.z.string().max(1000).optional(),
    anti_patterns: zod_1.z.string().max(500).optional(),
    documentation_url: zod_1.z.string().url().optional(),
    related_ledger_id: zod_1.z.string().uuid().optional(),
    project_id: zod_1.z.string().optional()
});
exports.UpdatePatternSchema = zod_1.z.object({
    name: zod_1.z.string().min(3).max(100).optional(),
    blessed_solution: zod_1.z.string().min(5).max(200).optional(),
    rationale: zod_1.z.string().max(1000).optional(),
    anti_patterns: zod_1.z.string().max(500).optional(),
    documentation_url: zod_1.z.string().url().optional(),
    status: zod_1.z.enum(['ACTIVE', 'DEPRECATED', 'EVALUATING']).optional()
});
// ============================================================================
// Protocol Schemas
// ============================================================================
exports.CreateProtocolSchema = zod_1.z.object({
    title: zod_1.z.string()
        .min(3, "Title must be at least 3 characters")
        .max(40, "Title must be at most 40 characters")
        .transform(s => s.toUpperCase()),
    description: zod_1.z.string()
        .min(10, "Description must be at least 10 characters")
        .max(120, "Description must be at most 120 characters"),
    content: zod_1.z.string()
        .min(20, "Content must be at least 20 characters")
        .max(10000, "Content must be at most 10000 characters"),
    project_id: zod_1.z.string().optional()
});
// ============================================================================
// Governance Request Schemas
// ============================================================================
exports.CreateGovernanceRequestSchema = zod_1.z.object({
    title: zod_1.z.string()
        .min(5, "Title must be at least 5 characters")
        .max(150, "Title must be at most 150 characters"),
    description: zod_1.z.string().max(2000).optional(),
    request_type: zod_1.z.enum(['FEATURE_APPROVAL', 'ARCHITECTURE_REVIEW', 'POLICY_QUESTION', 'EXCEPTION_REQUEST', 'TOOL_EVALUATION']),
    domain: zod_1.z.enum(['ARCHITECTURE', 'DATA', 'STANDARDS', 'SECURITY', 'STRATEGY']),
    urgency: zod_1.z.enum(['LOW', 'STANDARD', 'ELEVATED', 'CRITICAL']).default('STANDARD'),
    complexity: zod_1.z.enum(['TRIVIAL', 'SIMPLE', 'MODERATE', 'COMPLEX', 'EPIC']).default('MODERATE'),
    requester: zod_1.z.string().max(100).optional(),
    project_id: zod_1.z.string().optional()
});
exports.ResolveGovernanceRequestSchema = zod_1.z.object({
    resolution_summary: zod_1.z.string().min(10).max(500),
    resolution_ledger_id: zod_1.z.string().uuid().optional()
});
// ============================================================================
// Project Schemas
// ============================================================================
exports.CreateProjectSchema = zod_1.z.object({
    name: zod_1.z.string()
        .min(2, "Project name must be at least 2 characters")
        .max(50, "Project name must be at most 50 characters")
        .transform(s => s.toUpperCase().replace(/\s+/g, '_')),
    description: zod_1.z.string().max(500).optional()
});
