// @stackbilt/scaffold-core/dist/index.js
"use strict";
/**
 * @stackbilt/scaffold-core
 *
 * Zero-dependency, zero-inference, zero-network scaffold engine core.
 *
 * Entrypoint: buildScaffold(intention, options?) → LocalScaffoldResult
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.materializeScaffold = exports.addGovernanceFiles = exports.generateFiles = exports.buildGovernance = exports.getKnowledge = exports.classify = void 0;
exports.buildScaffold = buildScaffold;
exports.buildOracleContext = buildOracleContext;
// ============================================================================
// Sub-module public APIs
// ============================================================================
var index_1 = require("./classify/index");
Object.defineProperty(exports, "classify", { enumerable: true, get: function () { return index_1.classify; } });
var index_2 = require("./knowledge/index");
Object.defineProperty(exports, "getKnowledge", { enumerable: true, get: function () { return index_2.getKnowledge; } });
var index_3 = require("./governance/index");
Object.defineProperty(exports, "buildGovernance", { enumerable: true, get: function () { return index_3.buildGovernance; } });
var index_4 = require("./codegen/index");
Object.defineProperty(exports, "generateFiles", { enumerable: true, get: function () { return index_4.generateFiles; } });
Object.defineProperty(exports, "addGovernanceFiles", { enumerable: true, get: function () { return index_4.addGovernanceFiles; } });
var index_5 = require("./materializer/index");
Object.defineProperty(exports, "materializeScaffold", { enumerable: true, get: function () { return index_5.materializeScaffold; } });
const index_6 = require("./classify/index");
const index_7 = require("./knowledge/index");
const index_8 = require("./governance/index");
const index_9 = require("./codegen/index");
const index_10 = require("./materializer/index");
const bindings_1 = require("./classify/bindings");
/**
 * Build a complete scaffold result from a plain-English intention string.
 *
 * Orchestration order:
 *   1. classify(intention)        → ClassifyResult
 *   2. getKnowledge(pattern, ...) → PatternKnowledge
 *   3. buildGovernance(facts, ...) → GovernanceDocs
 *   4. generateFiles(facts)       → ScaffoldFile[] (base + routes)
 *   5. addGovernanceFiles(...)    → grafts .ai/*.md onto the file list
 *   6. materializeScaffold(facts) → ADF + project files (grafted in)
 *
 * @param intention - Plain-English description of what to build
 * @param options   - Optional overrides (projectName, oracle mode)
 * @returns         - LocalScaffoldResult with all scaffold artifacts
 */
function buildScaffold(intention, options = {}) {
    const classification = (0, index_6.classify)(intention);
    const knowledge = (0, index_7.getKnowledge)(classification.pattern, classification.qualityProfile.complianceDomains);
    const bindings = (0, bindings_1.inferBindings)(classification.pattern, classification.traits);
    const facts = {
        pattern: classification.pattern,
        projectName: options.projectName ?? 'my-worker',
        intention: classification.enrichedIntention,
        bindings,
        traits: classification.traits,
        qualityProfile: classification.qualityProfile,
    };
    const governance = (0, index_8.buildGovernance)(facts, knowledge);
    // Generate base + route files, then graft governance docs on top
    const codegenFiles = (0, index_9.generateFiles)(facts);
    const filesWithGovernance = (0, index_9.addGovernanceFiles)(codegenFiles, governance);
    // Materialize ADF + project files; graft only ADF/contract files not already present
    let finalFiles = filesWithGovernance;
    try {
        const { files: materializedFiles } = (0, index_10.materializeScaffold)(facts);
        const existingPaths = new Set(finalFiles.map((f) => f.path));
        for (const mf of materializedFiles) {
            const isAdf = mf.path.startsWith('.ai/') && mf.path.endsWith('.adf');
            const isContract = mf.path.startsWith('src/contracts/');
            const isSchema = mf.path === 'schema.sql';
            if ((isAdf || isContract || isSchema) && !existingPaths.has(mf.path)) {
                finalFiles = [...finalFiles, mf];
            }
        }
    }
    catch {
        // Materializer failure is non-fatal — codegen output is still complete
    }
    return {
        classification,
        knowledge,
        governance,
        files: finalFiles,
        facts,
        traits: classification.traits,
        tier2Recommended: classification.confidence < 0.6,
    };
}
/**
 * Derive oracle context from a LocalScaffoldResult for use by an LLM polish pass.
 *
 * All fields are derived from the existing result — no additional inference or
 * network calls required. Consumers pass this to the oracle instead of the
 * old promptContext field that was stripped when migrating from the local shim.
 *
 * @see stackbilt-web oracle.ts
 * @see charter#224
 */
function buildOracleContext(result) {
    const { classification, knowledge, governance, facts, files, tier2Recommended } = result;
    return {
        intention: facts.intention,
        pattern: classification.pattern,
        meta: {
            confidence: classification.confidence,
            tier2Recommended,
            testingLevel: classification.qualityProfile.testingLevel,
            complianceDomains: classification.qualityProfile.complianceDomains,
            observability: classification.qualityProfile.observability,
            authentication: classification.qualityProfile.authentication,
            rateLimiting: classification.qualityProfile.rateLimiting,
        },
        traits: classification.traits,
        runtime: {
            bindings: facts.bindings.map(b => ({ type: b.type, name: b.name, binding: b.binding })),
            piiHandling: classification.qualityProfile.piiHandling,
        },
        governance: {
            threatModel: governance.threatModel,
            adr001: governance.adr001,
            adr002: governance.adr002 ?? null,
            testPlan: governance.testPlan,
        },
        knowledge: {
            adrContext: knowledge.adrContext,
            adrDecision: knowledge.adrDecision,
            topThreats: knowledge.threats.slice(0, 5).map(t => ({
                id: t.id,
                description: t.description,
                mitigation: t.mitigation,
                severity: t.severity,
            })),
        },
        files: files.map(f => ({ path: f.path, content: f.content, role: f.role })),
    };
}
