// @stackbilt/scaffold-core/dist/materializer/adf.js
"use strict";
/**
 * materializer/adf — ADF file generators
 *
 * Generates .ai/core.adf, .ai/state.adf, .ai/manifest.adf for a scaffold output.
 * Extracted from stackbilt-web/src/lib/scaffold-materializer.ts.
 * Pure — no CF Worker imports, no network calls.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.str = str;
exports.slugify = slugify;
exports.traitsInclude = traitsInclude;
exports.renderManifestAdf = renderManifestAdf;
exports.renderCoreAdf = renderCoreAdf;
exports.renderStateAdf = renderStateAdf;
exports.generateAdfFiles = generateAdfFiles;
function str(facts, key, fallback = '') {
    const v = facts[key];
    if (typeof v === 'string')
        return v;
    if (Array.isArray(v))
        return v.join(', ');
    return fallback;
}
function slugify(name) {
    return name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
}
function traitsInclude(facts, ...terms) {
    const v = facts.runtime_traits;
    const joined = Array.isArray(v) ? v.join(' ') : typeof v === 'string' ? v : '';
    return terms.some(t => joined.includes(t));
}
// ─── Template renderers ───────────────────────────────────────────────────────
function renderManifestAdf(facts, projectName) {
    const confidence = str(facts, 'scaffold_confidence', 'moderate');
    const rawBalance = facts.elemental_balance;
    const balance = typeof rawBalance === 'object' && rawBalance !== null
        ? Object.entries(rawBalance).filter(([, v]) => v > 0).map(([k, v]) => `${k}:${v}`).join(' ') || 'neutral'
        : str(facts, 'elemental_balance', 'unknown');
    const shadowDensity = facts.shadow_density ?? 'unknown';
    return `# ${projectName} — ADF Manifest
# Generated by Stackbilt scaffold engine
# Confidence: ${confidence} | Shadow density: ${shadowDensity}

version: "0.1"
project: "${projectName}"

## Modules

- core.adf        # Product requirements, UX, security
- state.adf       # Sprint backlog and current task

## On-Demand Triggers

| Domain     | Trigger Keywords                     |
|------------|--------------------------------------|
| product    | ${str(facts, 'requirement_name')}, requirements, features |
| ux         | ${str(facts, 'interface_name')}, layout, components |
| security   | ${str(facts, 'threat_name')}, threat, mitigation |
| runtime    | ${str(facts, 'runtime_name')}, deploy, worker |
| testing    | ${str(facts, 'test_plan_name')}, test, coverage |
| sprint     | ${str(facts, 'first_task_name')}, task, estimate |

## Metrics

| Metric              | Value         |
|---------------------|---------------|
| position_count      | ${facts.position_count ?? 6} |
| shadow_density      | ${shadowDensity} |
| elemental_balance   | ${balance}    |
| scaffold_confidence | ${confidence} |
`;
}
function renderCoreAdf(facts, projectName) {
    const reqName = str(facts, 'requirement_name');
    const reqPriority = str(facts, 'requirement_priority', 'P1');
    const reqEffort = str(facts, 'requirement_effort', 'medium');
    const reqAcceptance = str(facts, 'requirement_acceptance');
    const ifaceName = str(facts, 'interface_name');
    const ifaceRegions = str(facts, 'interface_regions');
    const ifaceGrid = str(facts, 'interface_grid');
    const ifaceComponents = str(facts, 'interface_components');
    const threatName = str(facts, 'threat_name');
    const threatLikelihood = str(facts, 'threat_likelihood');
    const threatImpact = str(facts, 'threat_impact');
    const threatMitigation = str(facts, 'threat_mitigation');
    return `# ${projectName} — Core
# Product requirements, UX patterns, and security constraints

## Product Requirements

### ${reqName}
- **Priority**: ${reqPriority}
- **Effort**: ${reqEffort}
- **Acceptance criteria**: ${reqAcceptance}

## UX Pattern

### ${ifaceName}
- **Regions**: ${ifaceRegions}
- **Grid**: ${ifaceGrid}
- **Components**: ${ifaceComponents}

## Security

### ${threatName}
- **Likelihood**: ${threatLikelihood}
- **Impact**: ${threatImpact}
- **Mitigation**: ${threatMitigation}
`;
}
function renderStateAdf(facts, projectName) {
    const taskName = str(facts, 'first_task_name');
    const taskEstimate = str(facts, 'first_task_estimate');
    const taskComplexity = str(facts, 'first_task_complexity');
    const taskDeliverable = str(facts, 'first_task_deliverable');
    return `# ${projectName} — State
# Current sprint backlog

## Current Sprint

### ${taskName}
- **Estimate**: ${taskEstimate} points
- **Complexity**: ${taskComplexity}
- **Deliverable**: ${taskDeliverable}
- **Status**: not_started

## Velocity

| Sprint | Points Planned | Points Done |
|--------|---------------|-------------|
| 1      | ${taskEstimate}            | —           |
`;
}
// ─── Public API ───────────────────────────────────────────────────────────────
/**
 * Generate ADF ScaffoldFiles (.ai/manifest.adf, .ai/core.adf, .ai/state.adf)
 * for the given scaffold facts.
 *
 * @param scaffoldFacts - Structured scaffold facts from the classify pipeline
 * @param rawFacts      - Optional raw key-value facts from scaffold-cast (fallback to empty)
 */
function generateAdfFiles(scaffoldFacts, rawFacts = {}) {
    const projectName = scaffoldFacts.projectName;
    return [
        { path: '.ai/manifest.adf', content: renderManifestAdf(rawFacts, projectName), role: 'adf' },
        { path: '.ai/core.adf', content: renderCoreAdf(rawFacts, projectName), role: 'adf' },
        { path: '.ai/state.adf', content: renderStateAdf(rawFacts, projectName), role: 'adf' },
    ];
}
