// @stackbilt/scaffold-core/dist/classify/index.js
"use strict";
/**
 * classify — public API
 *
 * classify(intention) → ClassifyResult
 *
 * Thin wrapper that calls choosePattern, then patches the result with
 * inferQualityProfile and extractEnrichedPrd so callers get a fully-
 * populated ClassifyResult in one call.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.inferQualityProfile = exports.inferBindings = exports.DOMAIN_ENTITY_PATTERNS = exports.STACK_TOKENS = exports.injectPrdSections = exports.extractEnrichedPrd = exports.keywordScore = exports.choosePattern = exports.PATTERNS = void 0;
exports.classify = classify;
var patterns_1 = require("./patterns");
Object.defineProperty(exports, "PATTERNS", { enumerable: true, get: function () { return patterns_1.PATTERNS; } });
var classifier_1 = require("./classifier");
Object.defineProperty(exports, "choosePattern", { enumerable: true, get: function () { return classifier_1.choosePattern; } });
Object.defineProperty(exports, "keywordScore", { enumerable: true, get: function () { return classifier_1.keywordScore; } });
var enricher_1 = require("./enricher");
Object.defineProperty(exports, "extractEnrichedPrd", { enumerable: true, get: function () { return enricher_1.extractEnrichedPrd; } });
Object.defineProperty(exports, "injectPrdSections", { enumerable: true, get: function () { return enricher_1.injectPrdSections; } });
Object.defineProperty(exports, "STACK_TOKENS", { enumerable: true, get: function () { return enricher_1.STACK_TOKENS; } });
Object.defineProperty(exports, "DOMAIN_ENTITY_PATTERNS", { enumerable: true, get: function () { return enricher_1.DOMAIN_ENTITY_PATTERNS; } });
var bindings_1 = require("./bindings");
Object.defineProperty(exports, "inferBindings", { enumerable: true, get: function () { return bindings_1.inferBindings; } });
var quality_1 = require("./quality");
Object.defineProperty(exports, "inferQualityProfile", { enumerable: true, get: function () { return quality_1.inferQualityProfile; } });
const classifier_2 = require("./classifier");
const enricher_2 = require("./enricher");
const quality_2 = require("./quality");
const patterns_2 = require("./patterns");
/**
 * Classify an intention string and return a fully-populated ClassifyResult.
 *
 * Steps:
 * 1. choosePattern — vocabulary scoring → PatternName + traits
 * 2. extractEnrichedPrd — stack/entity detection
 * 3. injectPrdSections — append Stack/Entities sections to intention
 * 4. inferQualityProfile — compliance + feature flags
 */
function classify(intention) {
    const base = (0, classifier_2.choosePattern)(intention);
    // Resolve source_pattern string from the scored patterns for quality inference
    const matchedScored = patterns_2.SCORED_PATTERNS.find((p) => p.name === base.pattern && p.traits.every((t) => base.traits.includes(t)));
    const sourcePattern = matchedScored?.traitMap['source_pattern'] ?? base.pattern;
    const { stack, entities } = (0, enricher_2.extractEnrichedPrd)(intention);
    const enriched = (0, enricher_2.injectPrdSections)(intention, stack, entities);
    const qualityProfile = (0, quality_2.inferQualityProfile)(enriched, sourcePattern, base.traits);
    return {
        ...base,
        qualityProfile,
        enrichedIntention: enriched,
    };
}
