// @stackbilt/scaffold-core/dist/codegen/index.js
"use strict";
/**
 * codegen — public API
 *
 * generateFiles(facts) → ScaffoldFile[]
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateWranglerBindings = exports.addGovernanceFiles = exports.baseFiles = exports.routeContent = exports.buildRoutes = void 0;
exports.generateFiles = generateFiles;
var routes_1 = require("./routes");
Object.defineProperty(exports, "buildRoutes", { enumerable: true, get: function () { return routes_1.buildRoutes; } });
Object.defineProperty(exports, "routeContent", { enumerable: true, get: function () { return routes_1.routeContent; } });
var files_1 = require("./files");
Object.defineProperty(exports, "baseFiles", { enumerable: true, get: function () { return files_1.baseFiles; } });
Object.defineProperty(exports, "addGovernanceFiles", { enumerable: true, get: function () { return files_1.addGovernanceFiles; } });
var wrangler_1 = require("./wrangler");
Object.defineProperty(exports, "generateWranglerBindings", { enumerable: true, get: function () { return wrangler_1.generateWranglerBindings; } });
const files_2 = require("./files");
const routes_2 = require("./routes");
/**
 * Generate the full set of ScaffoldFiles for the given facts.
 *
 * Produces base infrastructure files (wrangler.toml, package.json, etc.)
 * plus route handler files for each route in the pattern's default_routes trait.
 *
 * Note: governance docs (.ai/*.md) are added by the orchestrator after calling
 * buildGovernance(). Use addGovernanceFiles() to graft them in.
 */
function generateFiles(facts) {
    return [...(0, files_2.baseFiles)(facts), ...(0, routes_2.buildRoutes)(facts)];
}
