// @stackbilt/scaffold-core/dist/knowledge/index.js
"use strict";
/**
 * knowledge — public API
 *
 * getKnowledge(pattern, complianceDomains?) → PatternKnowledge
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.adrDecisionByPattern = exports.adrContextByPattern = exports.domainThreats = exports.patternSpecificThreats = void 0;
exports.getKnowledge = getKnowledge;
var threats_1 = require("./threats");
Object.defineProperty(exports, "patternSpecificThreats", { enumerable: true, get: function () { return threats_1.patternSpecificThreats; } });
Object.defineProperty(exports, "domainThreats", { enumerable: true, get: function () { return threats_1.domainThreats; } });
var decisions_1 = require("./decisions");
Object.defineProperty(exports, "adrContextByPattern", { enumerable: true, get: function () { return decisions_1.adrContextByPattern; } });
Object.defineProperty(exports, "adrDecisionByPattern", { enumerable: true, get: function () { return decisions_1.adrDecisionByPattern; } });
const threats_2 = require("./threats");
const decisions_2 = require("./decisions");
/**
 * Retrieve all knowledge (threats + ADR fragments) for a given source pattern.
 *
 * @param pattern           Source pattern name (e.g. 'rest-api', 'stripe-webhook')
 * @param complianceDomains Optional compliance domains for domain-specific threats
 */
function getKnowledge(pattern, complianceDomains) {
    return {
        threats: (0, threats_2.patternSpecificThreats)(pattern),
        adrContext: (0, decisions_2.adrContextByPattern)(pattern),
        adrDecision: (0, decisions_2.adrDecisionByPattern)(pattern),
        domainThreats: complianceDomains ? (0, threats_2.domainThreats)(complianceDomains) : [],
    };
}
