// @stackbilt/scaffold-core/dist/governance/index.js
"use strict";
/**
 * governance — public API
 *
 * buildGovernance(facts, knowledge) → GovernanceDocs
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTestPlan = exports.buildAdr002 = exports.buildAdr001 = exports.buildThreatModel = void 0;
exports.buildGovernance = buildGovernance;
var threat_model_1 = require("./threat-model");
Object.defineProperty(exports, "buildThreatModel", { enumerable: true, get: function () { return threat_model_1.buildThreatModel; } });
var adr_1 = require("./adr");
Object.defineProperty(exports, "buildAdr001", { enumerable: true, get: function () { return adr_1.buildAdr001; } });
Object.defineProperty(exports, "buildAdr002", { enumerable: true, get: function () { return adr_1.buildAdr002; } });
var test_plan_1 = require("./test-plan");
Object.defineProperty(exports, "buildTestPlan", { enumerable: true, get: function () { return test_plan_1.buildTestPlan; } });
const threat_model_2 = require("./threat-model");
const adr_2 = require("./adr");
const test_plan_2 = require("./test-plan");
/**
 * Build all governance documents for a scaffold project.
 */
function buildGovernance(facts, knowledge) {
    return {
        threatModel: (0, threat_model_2.buildThreatModel)(facts, knowledge),
        adr001: (0, adr_2.buildAdr001)(facts, knowledge),
        adr002: (0, adr_2.buildAdr002)(facts),
        testPlan: (0, test_plan_2.buildTestPlan)(facts),
    };
}
